lib/puppet/configurer.rb

```diff
@@ -472,6 +433,25 @@ def find_functional_server
   end
   private :find_functional_server
 
+  def last_used_environment
+    if Puppet::FileSystem.exist?(Puppet[:lastrunfile])
+      begin
+        last_run_environment = Puppet::Util::Yaml.safe_load_file(Puppet[:lastrunfile])['version']['environment']
+      rescue
+        last_run_environment = nil
+      end
+    end
+    if Puppet::FileSystem.exist?(Puppet[:lastrunreport]) && !last_run_environment
```

Comment:
Recommend checking `!last_run_environment` first, so we avoid checking if `lastrunreport` exists in the happy path.